File: 1123-lowest-common-ancestor-of-deepest-leaves/1123-lowest-common-ancestor-of-deepest-leaves.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def lcaDeepestLeaves(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        
        maxDepth = 0
        lca = root

        def dfs(curNode, depth):
            nonlocal maxDepth
            nonlocal lca
            if not curNode:
                return depth-1

            leftDepth = dfs(curNode.left, depth+1)
            rightDepth = dfs(curNode.right, depth+1)

            if leftDepth > maxDepth or rightDepth > maxDepth:
                lca = curNode
            elif leftDepth == rightDepth == maxDepth:
                lca = curNode
            
            maxDepth = max(leftDepth, rightDepth, maxDepth)

            return max(leftDepth, rightDepth)

        dfs(root, 0)
        return lca
    

File: 1123-lowest-common-ancestor-of-deepest-leaves/1123-lowest-common-ancestor-of-deepest-leaves.cpp
class Solution {
public:
    TreeNode* helper(unordered_set<TreeNode*>&s, TreeNode *root){
        if(!root) return NULL;
        if(s.count(root))return root;
        TreeNode* l = helper(s, root->left);
        TreeNode* r = helper(s, root->right);
        if(l && r)return root;
        return l ? l : r;
    }
    TreeNode* lcaDeepestLeaves(TreeNode* root) { 
        unordered_set<TreeNode*>s;
        queue<TreeNode*>q;
        q.push(root);
        while(q.size()){
            int n = q.size();
            s.clear();
            while(n--){
                TreeNode* temp = q.front();
                q.pop();
                s.insert(temp);
                if(temp->left) q.push(temp->left);
                if(temp->right) q.push(temp->right);
            }
        }
        return helper(s, root);
    }
};